addons/mail/static/src/discuss/call/common/call_menu.js

```diff
@@ -1,14 +1,43 @@
-import { Component, useState } from "@odoo/owl";
+import { Component, useEffect, useState } from "@odoo/owl";
 
 import { registry } from "@web/core/registry";
 import { useService } from "@web/core/utils/hooks";
+import { callActionsRegistry } from "./call_actions";
 
 export class CallMenu extends Component {
     static props = [];
     static template = "discuss.CallMenu";
     setup() {
         super.setup();
         this.rtc = useState(useService("discuss.rtc"));
+        this.state = useState({
+            icon: undefined,
+        });
+        this.lastActions = {};
+        this.actionsStack = [];
+        useEffect(
+            (...actions) => {
+                for (const [key, isActive] of actions) {
+                    if (isActive !== this.lastActions[key]) {
+                        if (isActive) {
+                            if (!this.actionsStack.includes(key)) {
+                                this.actionsStack.unshift(key);
+                            }
+                        } else {
+                            this.actionsStack.splice(this.actionsStack.indexOf(key), 1);
+                        }
+                    }
+                }
+
+                this.state.icon = callActionsRegistry.get(this.actionsStack[0], undefined)?.icon;
+
+                this.lastActions = Object.fromEntries(actions);
+            },
+            () =>
+                callActionsRegistry
+                    .getEntries()
+                    .map(([key, action]) => [key, action.isActive(this.rtc.selfSession)])
+        );
```

Comment:
Just note that current implementation need the component to be mounted to observe change of action, so they are registered in stack.

Right now it's fine because the call menu is always visible in systray while the call is active, and no screen removes it... Or is it? Yes, some screens in backend have no systray, like the barcode app:
<img width="1278" alt="Screenshot 2024-07-12 at 17 36 19 (3)" src="https://github.com/user-attachments/assets/f371362f-b198-4644-83e4-e79a3bb747a8">

No, the systray is not behind. It's actually not mounted. So if the call state changes, the actions won't be registered until next mounting. And because of this, it won't follow the right ordering of transition (let's say, when opening barcode I share screen and then deafen. Which icon does it show? One will be correct, the inverse won't).


----------


So all in all, would be best to not rely on rendering to observe the transition. Instead of you can rely on storing the call actions in a owl reactive (e.g. in `this.store.callActions`, the store and discuss JS models are reactives), and you can observe change with `onChange()` reactive values. This will work without requiring rendering a component.

The `CallMenu` component would therefore just observe the `this.store.lastSelfCallAction` to put the right icon.